### ISSUE TITLE
Consistent underruns in audio output redirection

### ISSUE THREAD

**Issue by iskunk** [FreeRDP/FreeRDP#6276]
I am using xfreerdp 2.1.1, as packaged in the `freerdp2-x11 2.1.1+dfsg1-0ubuntu0.18.04.1` package on an Ubuntu Linux 18.04 system on amd64.

I am connecting to an x64-based Windows 10 Enterprise system, using this command line:
```
xfreerdp \
	/bpp:24 \
	/sec:nla \
	/size:1550x1150 \
	-grab-keyboard \
	/sound:sys:alsa,dev:sysdefault \
	/u:user@DOMAIN \
	/log-level:DEBUG \
	/v:hostname
```

The connection is via a fast broadband link with low enough latency for normal interactive SSH use and the like.

I cannot get the short chime that plays whenever the Windows volume control (on the taskbar) is adjusted to sound smoothly on the client side. There are consistent, repetitive pops, multiple per second, each accompanied by a
```
ALSA lib pcm.c:8306:(snd_pcm_recover) underrun occurred
```
warning message. I've confirmed that this occurs using both Pulseaudio and straight ALSA. (The above invocation is for the latter.)

**Comment by relft** [comment]
hi @iskunk 
similar buffer underruns on https://github.com/FreeRDP/FreeRDP/issues/6263
do you have sound from mic or choppy sound? what compile option have you used  from xfreerdp /buildconfig ?

thanks

**Comment by iskunk** [comment]
Hello @relft. I have not tested microphone redirection yet, since audio output is supposed to be the "easier" problem.

The FreeRDP installation I am using was not compiled by me; it is the Ubuntu package in the 18.04/bionic release. The build config is as follows:
```
$ xfreerdp /buildconfig
This is FreeRDP version 2.1.1 (n/a)
Build configuration: BUILD_TESTING=OFF BUILTIN_CHANNELS=ON HAVE_AIO_H=1 HAVE_EXECINFO_H=1 HAVE_FCNTL_H=1 HAVE_INTTYPES_H=1 HAVE_JOURNALD_H=TRUE HAVE_MATH_C99_LONG_DOUBLE=1 HAVE_POLL_H=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK=ON HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL= HAVE_SYSLOG_H=1 HAVE_SYS_EVENTFD_H=1 HAVE_SYS_FILIO_H= HAVE_SYS_MODEM_H= HAVE_SYS_SELECT_H=1 HAVE_SYS_SOCKIO_H= HAVE_SYS_STRTIO_H= HAVE_SYS_TIMERFD_H=1 HAVE_TM_GMTOFF=1 HAVE_UNISTD_H=1 HAVE_XI_TOUCH_CLASS=1 WITH_ALSA=ON WITH_CAIRO=ON WITH_CCACHE=ON WITH_CHANNELS=ON WITH_CLANG_FORMAT=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CUPS=ON WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_RDPDR=OFF WITH_DEBUG_RDPEI=OFF WITH_DEBUG_RDPGFX=OFF WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF_AVAILABLE=0 WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_CLIPRDR=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF WITH_DEBUG_XV=OFF WITH_DSP_EXPERIMENTAL=OFF WITH_EVENTFD_READ_WRITE=1 WITH_FAAC=OFF WITH_FAAD2=OFF WITH_FFMPEG=OFF WITH_GFX_H264=OFF WITH_GPROF=OFF WITH_GSM=OFF WITH_GSSAPI=OFF WITH_GSTREAMER_0_10=OFF WITH_GSTREAMER_1_0=ON WITH_ICU=ON WITH_IPP=OFF WITH_JPEG=ON WITH_LAME=OFF WITH_LIBRARY_VERSIONING=ON WITH_LIBSYSTEMD=ON WITH_MACAUDIO=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO_AVAILABLE=0 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_OPENCL=OFF WITH_OPENH264=OFF WITH_OPENSLES=OFF WITH_OPENSSL=ON WITH_OSS=ON WITH_PAM=ON WITH_PCSC=ON WITH_PROFILER=OFF WITH_PROXY=OFF WITH_PULSE=ON WITH_SAMPLE=OFF WITH_SANITIZE_ADDRESS=OFF WITH_SANITIZE_ADDRESS_AVAILABLE=1 WITH_SANITIZE_MEMORY=OFF WITH_SANITIZE_MEMORY_AVAILABLE=1 WITH_SANITIZE_THREAD=OFF WITH_SANITIZE_THREAD_AVAILABLE=1 WITH_SERVER=ON WITH_SERVER_CHANNELS=ON WITH_SERVER_INTERFACE=ON WITH_SHADOW=ON WITH_SMARTCARD_INSPECT=OFF WITH_SOXR=OFF WITH_SSE2=ON WITH_SWSCALE=OFF WITH_THIRD_PARTY=OFF WITH_VALGRIND_MEMCHECK=OFF WITH_VALGRIND_MEMCHECK_AVAILABLE=1 WITH_WAYLAND=ON WITH_WINPR_TOOLS=ON WITH_X11=ON WITH_X264=OFF WITH_XCURSOR=ON WITH_XDAMAGE=ON WITH_XEXT=ON WITH_XFIXES=ON WITH_XI=ON WITH_XINERAMA=ON WITH_XKBFILE=ON WITH_XRANDR=ON WITH_XRENDER=ON WITH_XSHM=ON WITH_XTEST=ON WITH_XV=ON WITH_ZLIB=ON
Build type:          RELWITHDEBINFO
CFLAGS:              -g -O2 -fdebug-prefix-map=/build/freerdp2-pfgvsb/freerdp2-2.1.1+dfsg1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -Wall -Wno-unused-result -Wno-unused-but-set-variable -Wno-deprecated-declarations -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -g -DWINPR_DLL
Compiler:            GNU, 7.5.0
Target architecture: x64
```

**Comment by akallabeth** [comment]
@iskunk @relft does it work better with just `/sound`?

**Comment by akallabeth** [comment]
for the record, this seems to be specific to `alsa` as it is quite sensitive to timing (and with network lags we can not guarantee timely delivery of new data)

**Comment by iskunk** [comment]
I tried just plain `/sound`, and I do get better behavior that way: each chime has maybe a single click/pop, rather than multiple, and not even every time. If I play a longer music file, then the sound does appear to stabilize after a couple seconds into playing correctly.

I normally don't use Pulseaudio, however. What is it about ALSA that makes it more prone to underruns? Is it just that hardware buffers are smaller than (software) sound-server buffers, and no other buffering is happening on the client?

How is audio buffering supposed to work on an RDP client, anyway? Should the client have e.g. 1-3 seconds of future audio stored, which can be overridden by out-of-band "pause" or "stop" messages? Or are the client/server supposed to negotiate things so that the client has as small a buffer as possible, without going into underrun territory?

(FYI, I encountered this issue while developing a server-side audio plugin for xrdp, so I am particularly interested in the "how this is supposed to work" question.)

**Comment by akallabeth** [comment]
@iskunk no expert on alsa, but the sound stuff is highly delicate:
On the RDP side it works like this:
1. You get samples from the server
1. you play the samples
1. you send back an ACK with proper timing
   * This is used by the server to manage the dataflow (adjust codec, sample rate, bitrater, ...)

On the backend side however it must be managed differently and alsa is very fragile if there are delayed samples (which happens quite often with network latency in between)
So, it is most likely necessary to insert 'silent' frames, but that is currently not done as we mainly use `pulse` and it is working `good enough`

**Comment by iskunk** [comment]
When you get sound samples from the server, are you supposed to get them *just* before they are played? Or can the server run ahead a bit, and send you audio e.g. 1 or 2 seconds in advance, to better work around variable latency? (It seems like if the user hits Stop/Pause, the server could send a SNDCLOSE to stop the audio immediately even if the client's buffer is full, but I don't know if that's how it's supposed to work.)

One thing I noticed that was different between 2.0.0 and 2.1.1 was [this commit](https://github.com/FreeRDP/FreeRDP/commit/25eb6b6cddb5d78058ecdb67a073cb7f6ffbe2b7). If the server tries to send more than a small amount of audio in advance, it now gets dropped as an "overrun." But isn't having a bigger buffer on the client desirable, to avoid underruns? Does the RDP protocol put a limit on how much audio should be buffered?

I'm also curious to know the difference between how Pulseaudio and ALSA are managed that makes PA work better. Abstractly, both have a buffer, and that buffer is getting populated the same way, so I don't understand why one would underrun where the other one wouldn't. Hypothetically, what implementation work would FreeRDP's ALSA support need for it to work as well?

(Apologies for my many questions; I'd just like to understand the problem, and this thread could serve as useful documentation for others as well.)

**Comment by akallabeth** [comment]
@iskunk bigger buffer would be great for non realtime audio, yes.
Sadly different server implementations (even the ones between windows 7 and 10 and their server equivalents) handle that case not very graceful, so no sound close :/
Even worse, with older versions you then have to wait several seconds until the sound stops on higher latency links.

As for the difference between `alsa` and `pulse`, it looks like pulse does some magic if there is no data to play whereas alsa gets really noisy. Could be because `pulse` is always keeping the device open (the sound server component) and alsa is more low level and directly accessing the device

**Comment by iskunk** [comment]
Is there a way to increase the size of FreeRDP's audio buffer? It seems like there is a tradeoff between a bigger buffer (fewer/no underruns, but you can't stop the sound immediately), and a smaller one (lower latency, immediate stop, better for games, but underruns are likely if your connection isn't rock-solid).

This is especially important if FreeRDP drops excess audio as an "overrun," because that then kills the easiest strategy for avoiding underruns.

(I do understand a bit why that change was made... when I was testing my custom server-side audio driver with 2.0.0, I ran `aplay` on a 30-second sound file on the server. `aplay` exited after about a second, and FreeRDP played the entire 30 seconds of audio from its buffer. That's a bit crazy. But 2.1.1 goes to the other extreme, where e.g. 0.25 sec of audio sent in advance gets dropped as an overrun.)

Okay, that is an understandable difference between ALSA and Pulseaudio---they both get underruns, but PA is better at "hiding" them. That suggests that the focus is better placed on avoiding underruns in the first place.

**Comment by akallabeth** [comment]
@iskunk currently not adjustable. Feel free to experiment.
We´re happy about any improvement there, but the most relevant changes are most likely in #5267 which allows freerdp to use simpler and adjustable threadding (and therefore can improve stuff running in parallel, like audio and video decoding)

**Comment by iskunk** [comment]
I don't think that change is going to affect this issue much, because this is just a matter of transferring and buffering audio, not resource contention on the client. (I tested by *just* playing a chime; I wasn't doing anything crazy with video or other high-bandwidth material.)

I'll do some experimentation, but let me run something by you:

I've found that the RDP server side can have quite a bit of control over how audio is buffered on the client, just by controlling the timing of the wave messages it sends out. (E.g. the server can send out a wave message 0.1 sec in advance, or 5 seconds in advance.) Adjusting the buffer size on the client is comparatively not as useful, because even if you have a big buffer there, the server might not give you enough wave data to fill it.

So IMO, the important thing is for the client to allow the server to buffer a lot of audio (e.g. 2 seconds) if it wants to, without going overboard (e.g. 30 seconds is clearly too much).

What I'm thinking is to adjust the overrun-detect code to allow an arbitrary minimum (probably 2 seconds) of audio to be buffered, before it stops dropping data in excess of that as an overrun. If that allows me to get smooth audio working, would that be an acceptable PR? That wouldn't go against any part of the RDP spec, or cause other issues?

**Comment by akallabeth** [comment]
@iskunk well, the problem is that a slight stall in decoding increases the response time for the ack messages (not TCP, but RDP protocol level) and then the server adjusts quite some stuff to these new conditions. Bringing that delay down will fix a lot of these glitches as we do not create a condition that stalls input processing.

As for your suggestion with the configurable delay buffer, we experimented a lot with the timestamps sent back to the server, but that is not really something portable (every windows version behaves slightly different)

**Comment by iskunk** [comment]
If anything, future work could implement dropping audio frames if there is a delay in processing, so the audio always plays when it is scheduled. IMO, it would be better to drop audio frames in that situation (like if heavy swapping causes the system to freeze for a second or two), rather than drop them due to a narrow overrun threshold.

For the configurable delay/buffer, I was actually thinking of that on the server side (specifically with xrdp). While I can't change what the Windows server does, as long as the spec allows it, I can try to put together something that allows xrdp audio to play solidly.

**Comment by amazingfate** [comment]
I have a rk3288 arm board. When I play a 1080p video, the audio pause sometimes and after playing for sometimes, the audio is behind the video. As @iskunk mentioned in [this commit](https://github.com/FreeRDP/FreeRDP/commit/25eb6b6cddb5d78058ecdb67a073cb7f6ffbe2b7), if I remove the function rdpsnd_detect_overrun, the audio will sync with video perfectly. CPU is not full loaded when playing so I don't think cpu is the bottle neck causing this issue.

**Comment by akallabeth** [comment]
@amazingfate ouch, @llyzs has done some thorough research there, is seems it really depends on the remote os version how good this syncs.
will need further investigation here.

**Comment by amazingfate** [comment]
> @amazingfate ouch, @llyzs has done some thorough research there, is seems it really depends on the remote os version how good this syncs.
> will need further investigation here.

Yeah, my workaround only works on windows 10 server. On windows 7 it doesn't work.

**Comment by MichaelJTokarev** [comment]
Coming back to this prob 2 years later.
Trying sound from windows10 and windows7. They behave a bit differently, but the result on freerdp side is similar: with `/sound:sys:alsa` the audio is always coughing at the very beginning of each sound (differently depending on windows version), but becomes good after the first ~1/4s. While with `/sound:sys:pulse`, it always works fine.

What I also see is that pulse delays sound start for a little bit (maybe 1/4s), while alsa tries to output sound immediately.

On windows side there's "Remote audio Volume control" thing where one can click with mouse to change volume, and on each click, windows is playing the default sound. I'm "measuring" time from this click to the time when actual sound is heard. With pulse there's always a very small delay between the click and start of the sound, and whole sound is good. While with alsa, it starts playing immediately but becomes noisy within first little moment.

In both cases, further sound is always good, even over relatively slow remote links.

I dunno if the OP is having problems with sound "coughing" down the line or at the very beginning only, but for me it is always at the beginning only not counting things like network disconnect ofcourse).

freerdp 2.9.0.
